util/topsql/reporter/reporter.go

```diff
@@ -68,15 +70,173 @@ type cpuData struct {
 // dataPoints represents the cumulative SQL plan CPU time in current minute window
 // dataPoints do not guarantee the TimestampList is sorted by timestamp when there is a time jump backward.
 type dataPoints struct {
-	SQLDigest      []byte
-	PlanDigest     []byte
-	TimestampList  []uint64
-	CPUTimeMsList  []uint32
-	CPUTimeMsTotal uint64
+	SQLDigest             []byte
+	PlanDigest            []byte
+	TimestampList         []uint64
+	CPUTimeMsList         []uint32
+	CPUTimeMsTotal        uint64
+	StmtExecCountList     []uint64
+	StmtKvExecCountList   []map[string]uint64
+	StmtDurationSumNsList []uint64
+
+	// tsIndex is used to quickly find the corresponding array index through timestamp.
+	//
+	// map: timestamp => index of TimestampList / CPUTimeMsList / StmtExecCountList / ...
+	tsIndex map[uint64]int
+}
+
+func newDataPoints(sqlDigest, planDigest []byte) *dataPoints {
+	listCap := int(variable.TopSQLVariable.ReportIntervalSeconds.Load()/variable.TopSQLVariable.PrecisionSeconds.Load() + 1)
+	if listCap < 1 {
+		listCap = 1
+	}
+	return &dataPoints{
+		SQLDigest:             sqlDigest,
+		PlanDigest:            planDigest,
+		CPUTimeMsList:         make([]uint32, 0, listCap),
+		TimestampList:         make([]uint64, 0, listCap),
+		StmtExecCountList:     make([]uint64, 0, listCap),
+		StmtKvExecCountList:   make([]map[string]uint64, 0, listCap),
+		StmtDurationSumNsList: make([]uint64, 0, listCap),
+		tsIndex:               make(map[uint64]int, listCap),
+	}
+}
+
+func (d *dataPoints) rebuildTsIndex() {
+	if len(d.TimestampList) == 0 {
+		d.tsIndex = map[uint64]int{}
+		return
+	}
+	d.tsIndex = make(map[uint64]int, len(d.TimestampList))
+	for index, ts := range d.TimestampList {
+		d.tsIndex[ts] = index
+	}
+}
+
+func (d *dataPoints) appendCPUTime(timestamp uint64, cpuTimeMs uint32) {
```

Comment:
Does this mean that the dataPoints will not be ordered? If then, will it be simpler if dataPoints is a `map[ts]Metrics`?